go/extra/stats/cmd/stats.go

```diff
@@ -248,74 +282,84 @@ func getStats(ctx context.Context, consensus consensusAPI.ClientBackend, registr
 		}
 
 		// Get block.
-		block, err := consensus.GetBlock(ctx, height)
-		if err != nil {
-			logger.Error("failed to query block",
-				"err", err,
-				"height", height,
-			)
-			os.Exit(1)
-		}
-		var tmBlockMeta tmApi.BlockMeta
-		if err := cbor.Unmarshal(block.Meta, &tmBlockMeta); err != nil {
-			logger.Error("unmarshal error",
-				"meta", block.Meta,
-				"err", err,
-			)
-			os.Exit(1)
-		}
-
-		// Go over all signatures for a block.
-		for _, sig := range tmBlockMeta.LastCommit.Signatures {
-			if sig.Absent() {
-				logger.Debug("skipping absent signature")
-				continue
-			}
-			nodeTmAddr := sig.ValidatorAddress.String()
-
-			// Signatures are for previous height.
-			if err := ensureNodeTracking(ctx, stats, nodeTmAddr, height-1, registry); err != nil {
-				logger.Error("failed to query registry",
+		tmBlockMeta := getTmBlockMetaOrExit(ctx, consensus, height)
+
+		// Process the commit that is put on chain in this block.
+		if height > 1 {
+			// Commit is for previous height.
+			lastCommitHeight := height - 1
+
+			// Get validators.
+			// Hypothesis: this gets the validator set with priorities after they're incremented to get the round 0
+			// proposer.
+			vs, err := consensus.GetValidatorSet(ctx, lastCommitHeight)
+			if err != nil {
+				logger.Error("failed to query validators",
 					"err", err,
-					"height", height-1,
+					"height", lastCommitHeight,
 				)
 				os.Exit(1)
 			}
-
-			// Add signatures.
-			if err := stats.addNodeSignature(nodeTmAddr); err != nil {
-				logger.Error("failure adding signature",
+			var vals tmtypes.ValidatorSet
+			if err := aminoCodec.UnmarshalBinaryBare(vs.Meta, &vals); err != nil {
+				logger.Error("unmarshal validator set error",
+					"meta", vs.Meta,
 					"err", err,
+					"height", lastCommitHeight,
 				)
 				os.Exit(1)
 			}
-		}
 
-		// Add proposer sum (previous proposer).
-		if previousProposerAddr != "" {
-			// Only count round 0 proposals.
-			if tmBlockMeta.LastCommit.Round == 0 {
-				// Proposers are for previous height.
-				if err := ensureNodeTracking(ctx, stats, previousProposerAddr, height-1, registry); err != nil {
-					logger.Error("failed to query registry",
-						"err", err,
-						"height", height-1,
-					)
-					os.Exit(1)
-				}
+			// Go over all validators.
+			for _, val := range vals.Validators {
+				stats.nodeStatsOrExit(ctx, registry, lastCommitHeight, val.Address.String()).elections++
+			}
 
-				// Add proposal.
-				if err := stats.addNodeProposal(previousProposerAddr); err != nil {
-					logger.Error("failure adding proposal",
-						"err", err,
-					)
-					os.Exit(1)
+			// Go over all signatures for a block.
+			for _, sig := range tmBlockMeta.LastCommit.Signatures {
+				if sig.Absent() {
```

Comment:
While we're here, what about signatures for nil? Are those recorded too (I believe with the Tendermint upgrade they are) and should they be treated as "not signing"?

Reply:
don't know

Reply:
If you voted for `nil` in a round that succeeded, you failed to sign the block that was committed. Is this any more useful that not signing?

Reply:
it's an online conscientious objection

Reply:
I'll make it not count. only popular answer wins ):